lib/galaxy/model/__init__.py

```diff
@@ -6072,9 +6074,12 @@ def to_history_dataset_association(self, target_history, parent_id=None, add_to_
         sa_session.add(hda)
         hda.metadata = self.metadata
         if add_to_history and target_history:
-            target_history.add_dataset(hda)
-        with transaction(sa_session):
-            sa_session.commit()
+            target_history.stage_addition(hda)
+            if flush:
+                target_history.add_pending_items()
```

Comment:
```suggestion
                target_history.add_pending_items()
                flush = False
```
you know your committing in there, so no need to do again on the next line.

Can you check where this method is used and see if not adding the dataset to a history and the flush=True ever make sense ?

Reply:
The flush argument for the `add_datasets` function is always set to `False` in the current context, which means it won't trigger a commit. Right?
Also, I noticed that in other places, the datasets are committed after using this function, for example, in [this part of the code](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/hdas.py#L229-L235).

Reply:
Ah, right, I keep forgetting that. It's fine then. The second part still applies, can you check in which cases we do in fact want to commit ?

Reply:
I didn't get what should look for exactly.

Reply:
Flushing is different from committing: `session.commit()` will write to the database AND commit the current transaction, whereas `session.flush()` will write to the database transaction buffer without committing. What that means is that if we flush after adding or modifying a dataset, and on a subsequent commit there's any kind of db error that causes a rollback, whatever we flushed will be rolled back. If we commit, then the rollback on a subsequent commit won't affect the changes we have committed previously.

We have `flush` parameters in our code for historical reasons: pre-2.0, SQLAlchemy would commit on flush (with our `autocommit=True` setting), which is why the terms were used interchangeably. Now it's no longer the case; so my suggestion would be to use argument names consistent with what we do in the function (i.e., flush vs commit).

And with that said, I am not at all sure we should commit in that method :) The code @arash77 referenced replaced the pre-2.0 `session.flush`: we did that to ensure there was no behavioral change (a commit stayed a commit). Otherwise with hundreds of `session.flush` lines no longer committing, we'd never know what broke when moving to 2.0. HOWEVER, now that we *have* upgraded, we can take a hard look at all those commits and consider if we really need them, and, where possible, replace commits with flushes (unless we're optimizing like here, in which case we still can consider whether it's a commit that we need and not a flush).

Also, the code you referenced uses the `transaction` context manager to ensure that if there is no active transaction, we begin one before trying to commit. If there is a preceding `session.add` statement, there's guaranteed to be an active transaction.

Reply:
Thanks for the explanation! Just to clarify, for this case, should I change `flush` to `commit`?

Reply:
> Thanks for the explanation! Just to clarify, for this case, should I change `flush` to `commit`?

Assuming you mean the name of the argument, if you intend to commit then yes, please. If you think flushing might be sufficient (so no `session.commit()`, then keep it as flush. In other words, as long as the argument is consistent with the action.

Reply:
Yes, I intend to commit here.
Thank you, then I will just rename the argument.